libkbfs/bserver_memory.go

```diff
@@ -0,0 +1,374 @@
+// storing blocks in memory.
+type BlockServerMemory struct {
+	crypto Crypto
+	log    logger.Logger
+
+	lock sync.RWMutex
+	// m is nil after Shutdown() is called.
+	m map[BlockID]blockMemEntry
+}
+
+var _ BlockServer = (*BlockServerMemory)(nil)
+
+// NewBlockServerMemory constructs a new BlockServerMemory that stores
+// its data in memory.
+func NewBlockServerMemory(config Config) *BlockServerMemory {
+	return &BlockServerMemory{
+		config.Crypto(),
+		config.MakeLogger("BSM"),
+		sync.RWMutex{},
+		make(map[BlockID]blockMemEntry),
+	}
+}
+
+var errBlockServerMemoryShutdown = errors.New("BlockServerMemory is shutdown")
+
+// Get implements the BlockServer interface for BlockServerMemory
+func (b *BlockServerMemory) Get(ctx context.Context, id BlockID, tlfID TlfID,
+	context BlockContext) ([]byte, BlockCryptKeyServerHalf, error) {
+	b.log.CDebugf(ctx, "BlockServerMemory.Get id=%s tlfID=%s context=%s",
+		id, tlfID, context)
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	if b.m == nil {
+		return nil, BlockCryptKeyServerHalf{}, errBlockServerMemoryShutdown
+	}
+
+	entry, ok := b.m[id]
+	if !ok {
+		return nil, BlockCryptKeyServerHalf{}, BServerErrorBlockNonExistent{}
+	}
+
+	if entry.tlfID != tlfID {
+		return nil, BlockCryptKeyServerHalf{},
+			fmt.Errorf("TLF ID mismatch: expected %s, got %s",
+				entry.tlfID, tlfID)
+	}
+
+	refEntry, ok := entry.refs[context.GetRefNonce()]
+	if refEntry.Context != context {
+		return nil, BlockCryptKeyServerHalf{},
+			fmt.Errorf("Context mismatch: expected %s, got %s",
+				refEntry.Context, context)
+	}
+
+	return entry.blockData, entry.keyServerHalf, nil
+}
+
+func validateBlockServerPut(
+	crypto Crypto, id BlockID, context BlockContext, buf []byte) error {
+	if context.GetCreator() != context.GetWriter() {
+		return fmt.Errorf("Can't Put() a block with creator=%s != writer=%s",
+			context.GetCreator(), context.GetWriter())
+	}
+
+	if context.GetRefNonce() != zeroBlockRefNonce {
+		return fmt.Errorf("Can't Put() a block with a non-zero refnonce.")
+	}
+
+	bufID, err := crypto.MakePermanentBlockID(buf)
+	if err != nil {
+		return err
+	}
+
+	if bufID != id {
+		return fmt.Errorf(
+			"Block ID mismatch: expected %s, got %s", bufID, id)
+	}
+
+	return nil
+}
+
+// Put implements the BlockServer interface for BlockServerMemory
+func (b *BlockServerMemory) Put(ctx context.Context, id BlockID, tlfID TlfID,
+	context BlockContext, buf []byte,
+	serverHalf BlockCryptKeyServerHalf) error {
+	b.log.CDebugf(ctx, "BlockServerMemory.Put id=%s tlfID=%s context=%s",
+		id, tlfID, context)
+
+	err := validateBlockServerPut(b.crypto, id, context, buf)
+	if err != nil {
+		return err
+	}
+
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	if b.m == nil {
+		return errBlockServerMemoryShutdown
+	}
+
+	if entry, ok := b.m[id]; ok {
+		if entry.tlfID != tlfID {
+			return fmt.Errorf(
+				"TLF ID mismatch: expected %s, got %s",
+				entry.tlfID, tlfID)
+		}
+
+		// TODO: Is it okay for the creator to change?
```

Comment:
I don't think so.  Puts should really only ever happen once, though we should handle duplicate Puts idempotently.